CalibTracker/SiPixelESProducers/src/SiPixelGainCalibrationForHLTGPU.cc

```diff
@@ -0,0 +1,100 @@
+        dus[geom.offsetDU(GeomDetEnumerators::tkDetEnum[i])]->type().isTrackerStrip()) {
+      if(geom.offsetDU(GeomDetEnumerators::tkDetEnum[i]) < m_detectors) m_detectors = geom.offsetDU(GeomDetEnumerators::tkDetEnum[i]);
+    }
+  }
+
+  /*
+  std::cout << "caching calibs for " << m_detectors << " pixel detectors of size " << gains.data().size() << std::endl;
+  std::cout << "sizes " << sizeof(char) << ' ' << sizeof(uint8_t) << ' ' << sizeof(SiPixelGainForHLTonGPU::DecodingStructure) << std::endl;
+  */
+
+  cudaCheck(cudaMallocHost((void**) & gainForHLTonHost_, sizeof(SiPixelGainForHLTonGPU)));
+  //gainForHLTonHost_->v_pedestals = gainDataOnGPU_; // how to do this?
+
+  // do not read back from the (possibly write-combined) memory buffer
+  auto minPed  = gains.getPedLow();
+  auto maxPed  = gains.getPedHigh();
+  auto minGain = gains.getGainLow();
+  auto maxGain = gains.getGainHigh();
+  auto nBinsToUseForEncoding = 253;
+
+  // we will simplify later (not everything is needed....)
+  gainForHLTonHost_->minPed_ = minPed;
+  gainForHLTonHost_->maxPed_ = maxPed;
+  gainForHLTonHost_->minGain_= minGain;
+  gainForHLTonHost_->maxGain_= maxGain;
+
+  gainForHLTonHost_->numberOfRowsAveragedOver_ = 80;
+  gainForHLTonHost_->nBinsToUseForEncoding_    = nBinsToUseForEncoding;
+  gainForHLTonHost_->deadFlag_                 = 255;
+  gainForHLTonHost_->noisyFlag_                = 254;
+
+  gainForHLTonHost_->pedPrecision  = static_cast<float>(maxPed - minPed) / nBinsToUseForEncoding;
+  gainForHLTonHost_->gainPrecision = static_cast<float>(maxGain - minGain) / nBinsToUseForEncoding;
+
+  /*
+  std::cout << "precisions g " << gainForHLTonHost_->pedPrecision << ' ' << gainForHLTonHost_->gainPrecision << std::endl;
+  */
+
+  // fill the index map
+  auto const & ind = gains.getIndexes();
+  /*
+  std::cout << ind.size() << " " << m_detectors << std::endl;
+  */
+
+  for (auto i=0U; i<m_detectors; ++i) {
+    auto p = std::lower_bound(ind.begin(),ind.end(),dus[i]->geographicalId().rawId(),SiPixelGainCalibrationForHLT::StrictWeakOrdering());
+    assert (p!=ind.end() && p->detid==dus[i]->geographicalId());
+    assert(p->iend<=gains.data().size());
+    assert(p->iend>=p->ibegin);
+    assert(0==p->ibegin%2);
+    assert(0==p->iend%2);
+    assert(p->ibegin!=p->iend);
+    assert(p->ncols>0);
+    gainForHLTonHost_->rangeAndCols[i] = std::make_pair(SiPixelGainForHLTonGPU::Range(p->ibegin,p->iend), p->ncols);
+    // if (ind[i].detid!=dus[i]->geographicalId()) std::cout << ind[i].detid<<"!="<<dus[i]->geographicalId() << std::endl;
+    // gainForHLTonHost_->rangeAndCols[i] = std::make_pair(SiPixelGainForHLTonGPU::Range(ind[i].ibegin,ind[i].iend), ind[i].ncols);
+  }
+
+}
+
+SiPixelGainCalibrationForHLTGPU::~SiPixelGainCalibrationForHLTGPU() {
+  cudaCheck(cudaFreeHost(gainForHLTonHost_));
+}
+
+SiPixelGainCalibrationForHLTGPU::GPUData::~GPUData() {
+  if(gainForHLTonGPU != nullptr) {
```

Comment:
You can avoid the check, as `cudaFree` will not do anything if the argument is a null pointer